drivers/ethernet/eth_stm32_hal.c

```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
 ETH_NET_DEVICE_DT_INST_DEFINE(0, eth_initialize,
 		    NULL, &eth0_data, &eth0_config,
 		    CONFIG_ETH_INIT_PRIORITY, &eth_api, ETH_STM32_HAL_MTU);
+
+#if defined(CONFIG_PTP_CLOCK_STM32_HAL)
+
+struct ptp_context {
+	struct eth_stm32_hal_dev_data *eth_dev_data;
+};
+
+static struct ptp_context ptp_stm32_0_context;
+
+static int ptp_clock_stm32_set(const struct device *dev,
+			      struct net_ptp_time *tm)
+{
+	struct ptp_context *ptp_context = dev->data;
+	struct eth_stm32_hal_dev_data *eth_dev_data = ptp_context->eth_dev_data;
+	ETH_HandleTypeDef *heth = &eth_dev_data->heth;
+	int key;
+
+	key = irq_lock();
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACSTSUR = tm->second;
+	heth->Instance->MACSTNUR = tm->nanosecond;
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSINIT;
+	while (heth->Instance->MACTSCR & ETH_MACTSCR_TSINIT_Msk) {
+		/* spin lock */
+	}
+#else
+	heth->Instance->PTPTSHUR = tm->second;
+	heth->Instance->PTPTSLUR = tm->nanosecond;
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSSTI;
+	while (heth->Instance->PTPTSCR & ETH_PTPTSCR_TSSTI_Msk) {
+		/* spin lock */
+	}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int ptp_clock_stm32_get(const struct device *dev,
+			      struct net_ptp_time *tm)
+{
+	struct ptp_context *ptp_context = dev->data;
+	struct eth_stm32_hal_dev_data *eth_dev_data = ptp_context->eth_dev_data;
+	ETH_HandleTypeDef *heth = &eth_dev_data->heth;
+	int key;
+
+	key = irq_lock();
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	tm->second = heth->Instance->MACSTSR;
+	tm->nanosecond = heth->Instance->MACSTNR;
+#else
+	tm->second = heth->Instance->PTPTSHR;
+	tm->nanosecond = heth->Instance->PTPTSLR;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int ptp_clock_stm32_adjust(const struct device *dev, int increment)
+{
+	struct ptp_context *ptp_context = dev->data;
+	struct eth_stm32_hal_dev_data *eth_dev_data = ptp_context->eth_dev_data;
+	ETH_HandleTypeDef *heth = &eth_dev_data->heth;
+	int key, ret;
+
+	if ((increment <= (int32_t)-NSEC_PER_SEC) ||
+			(increment >= (int32_t)NSEC_PER_SEC)) {
+		ret = -EINVAL;
+	} else {
+		key = irq_lock();
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+		heth->Instance->MACSTSUR = 0;
+		if (increment >= 0) {
+			heth->Instance->MACSTNUR = increment;
+		} else {
+			heth->Instance->MACSTNUR = ETH_MACSTNUR_ADDSUB | (NSEC_PER_SEC + increment);
+		}
+		heth->Instance->MACTSCR |= ETH_MACTSCR_TSUPDT;
+		while (heth->Instance->MACTSCR & ETH_MACTSCR_TSUPDT_Msk) {
+			/* spin lock */
```

Comment:
Depending on function's call place, it could be potentially dangerous. Maybe k_yield() needed here? Bu I am also not sure.

Reply:
Not sure if that's a great ideas since IRQs are locked in this context, would yielding back to scheduler without unlocking the irq actually be an issue ?

Reply:
I wanted to use irq_lock() here and particularly in `ptp_clock_stm32_get` because we use it in a zero-latency callback which doesn't allow any kernel code.  Basically, getting time should be as atomic operation as possible.

Reply:
Ok.